odoo/api.py

```diff
@@ -765,22 +798,16 @@ def protecting(self, what, records=None):
         - ``what`` a collection of pairs ``(fields, records)``.
         """
         protected = self._protected
+        top = None
         try:
-            protected.pushmap()
-            if records is not None:  # Handle first signature
-                ids_by_field = {field: records._ids for field in what}
-            else:  # Handle second signature
-                ids_by_field = defaultdict(list)
-                for fields, what_records in what:
-                    for field in fields:
-                        ids_by_field[field].extend(what_records._ids)
-
-            for field, rec_ids in ids_by_field.items():
-                ids = protected.get(field)
-                protected[field] = ids.union(rec_ids) if ids else frozenset(rec_ids)
+            top = protected.pushmap()
+            self._protect(what, records)
             yield
         finally:
-            protected.popmap()
+            assert protected.popmap() is top
+
+    def protector(self):
```

Comment:
This is overkill.
You can use `with env.protecting(...)` or a `ExitStack` if you need to enter at some point in your block.

Reply:
Not really, because we get the information of what needs to be protected only at the end of the stack, so we need to communicate that to the context manager.

Note that the context manager is used only to be 100% sure that we never forget to pop the stack

Reply:
I have added an example using `ExitStack` that does what you need. You can undo all changes in this file.